spark/src/main/scala/geotrellis/spark/io/FilteringLayerReader.scala

```diff
@@ -1,17 +1,26 @@
 package geotrellis.spark.io
 
 import geotrellis.spark.Boundable
+import geotrellis.spark.io.index.KeyIndex
+import geotrellis.spark.io.json._
+import spray.json.JsonFormat
 
-abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, ReturnType] {
+abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, K, ReturnType] {
 
-  def read(id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int): ReturnType
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int, format: JsonFormat[I]): ReturnType
+
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, numPartitions: Int, format: JsonFormat[I]): ReturnType =
```

Comment:
I'm not crazy about the idea of providing the format per read instruction, its one of those cases where you couple code structure to data structure. 

Perhaps we should consider making the `JsonFormat[KeyIndex[K]]` extensible. That class already has to partially read and match some some tag to figure out what parser to use. Why not expose that to the user and allow them to bring that in through higher precedence import. Something like:

```scala
class ChainedJsonFormat[T](
  base: JsonFormat[T], 
  fread: PartialFunction[JsValue,T], 
  fwrite PartialFunction[T,JsValue]
) extends JsonFormat[T] {
  def read(json: JsValue) = fread.applyOrElse(base.read(json))
  def write(thing: T) = fwrite.applyOrElse(base.write(thing))
}

object package spark
   implicit class withExtentFormat[T](fmt: JsonFormat[T]) {
     def extendFormat(fread: PartialFunction[JsValue,T], fwrite PartialFunction[T,JsValue]) =
		new ChainedJsonFormat(fmt, fread, fwrite)  
   }
}

object SomeUserModule {
  implicit val myProjectKeyFormat: JsonFormat[BoundKeyIndex[K]] = geotrellis.spark.json.baseKeyIndexFormat.extendFormat {
    json => {//parse me some json for y KeyIndex},
    thing => {// write it out using my rules}
  }

}
```